### ISSUE TITLE
How operator TopK should handle nan value?

### ISSUE THREAD

**Issue by xadupre** [onnx/onnx#4716]
# Ask a Question
TopK specifications does not specify how to handle nan values. onnxruntime seems to consider them as the top values. Should we ignore them?

**Comment by robertknight** [comment]
Several other operators that sort floats are also missing details of how NaNs should be handled:

- ArgMin
- ArgMax
- Min
- Max
- ReduceMin
- ReduceMax

For comparison, the `numpy.{min, max, argmin, argmax, minimum, maximum}` functions propagate NaNs. There are also variants of these (eg. `numpy.nanmin` that ignore NaNs.

**Comment by robertknight** [comment]
For reference, PyTorch 2.0's `topk` method considers NaN values greater than any other. This applies whether `largest` is True or False.

```
>>> x = torch.tensor([1., 2., -3., 4., float('nan'), float('inf')])
>>> x.topk(6, sorted=True)
torch.return_types.topk(
values=tensor([nan, inf,  4.,  2.,  1., -3.]),
indices=tensor([4, 5, 3, 1, 0, 2]))

>>> x.topk(6, sorted=True, largest=False)
torch.return_types.topk(
values=tensor([-3.,  1.,  2.,  4., inf, nan]),
indices=tensor([2, 0, 1, 3, 5, 4]))
```

Using NumPy's [np.partition](https://seanlaw.github.io/2020/01/03/finding-top-or-bottom-k-in-a-numpy-array/) to implement top-k produces the same result.

In the case where `largest=True`, treating NaNs as top values is generally consistent with other reduction operators that propagate NaNs. I suppose when `largest=False` there is a choice of either propagating NaNs (by treating them as being smaller than other values) or being consistent with the `largest=True` ordering.

**Comment by Raghul-M** [comment]
@justinchuby Is this still open? If yes, I’d like to work on it

**Comment by cbourjau** [comment]
Please note that the IEEE-754 actually defines a total ordering for floating point values: https://en.wikipedia.org/wiki/IEEE_754

**Comment by justinchuby** [comment]
This is still open. Contribution welcomed

**Comment by ogulcanaydogan** [comment]
Hi! I'd like to work on this. Based on the discussion:

1. **Proposed NaN handling for TopK:** Follow IEEE 754 total ordering, treating NaN as greater than all other values (consistent with PyTorch's `topk` behavior as @robertknight noted)
2. **Scope:** Update the TopK operator spec to explicitly define NaN behavior, and fix the reference implementation to match
3. **Related operators:** Could also address the same NaN ambiguity in ArgMin/ArgMax/Min/Max/ReduceMin/ReduceMax as a follow-up

I'll start with TopK as the primary fix and can extend to the related operators if maintainers prefer a unified approach.

Should this be a spec clarification in the current opset or a new opset version?

**Comment by justinchuby** [comment]
A spec clarification sounds reasonable. Thanks.

**Comment by LeonxLJX** [comment]
I'd like to take this one.

From the report, the useful first step is to pin down the exact code path involved before touching anything — these tend to have a small true cause behind a broad-looking symptom, and guessing at it usually produces a patch that masks the behavior instead of fixing it.

Plan: trace the path, reproduce it, then make a focused change with a test if this area has coverage. I'll keep the diff small so it's easy to review.

If there's an existing convention in this repo for how this kind of thing is handled, I'd rather follow it than invent my own — happy to be pointed at one.

(claiming via @LeonxLJX)
